[llvm][utils] Add a script to use PRs over pushing to main

ilovepi opened the pull request:
As discussed on discourse
https://discourse.llvm.org/t/rfc-require-pull-requests-for-all-llvm-project-commits/88164

This is an attempt at a script to automatically make and land PRs.

It creates a branch (or a series of branches for stacks), and lands them one by one, without waiting for review or checks to pass.

It supports --auto-merge, for single PRs. For stacks, use Graphite or some other tool.

It can work with GitHub tokens for https or ssh keys.

Example:
```console
GITHUB_TOKEN=$(gh auth token) python3 llvm_push_pr.py --upstream-remote origin
```

ilovepi committed:
[llvm][utils] Add a script to use PRs over pushing to main

As discussed on discourse
https://discourse.llvm.org/t/rfc-require-pull-requests-for-all-llvm-project-commits/88164

This is an attempt at a script to automatically make and land PRs.

It creates a branch (or a series of branches for stacks), and lands them
one by one, without waiting for review or checks to pass.

It supports --auto-merge, for single PRs. For stacks, use Graphite or
some other tool.

It can work with GitHub tokens for https or ssh keys.

Example:
```console
GITHUB_TOKEN=$(gh auth token) python3 llvm_push_pr.py --upstream-remote origin
```

boomanaiden154 commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+import time
+from typing import List, Optional
+
+import requests
```
Is there a reason to prefer `requests` (third party library) over the stdlib `urllib.request`?

ilovepi replied:
Moved to use just things from the sandard lib.

boomanaiden154 commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+    def __init__(
+        self, dry_run: bool = False, verbose: bool = False, quiet: bool = False
+    ):
+        """Initializes the Printer with dry_run, verbose, and quiet settings."""
```
This docstring seems pretty redundant. They're also not required by most style guides (including the Google one) for short functions.

ilovepi replied:
done.

boomanaiden154 commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+        self.quiet = quiet
+
+    def print(self, message: str, file=sys.stdout):
+        """Prints a message to the specified file, respecting quiet mode."""
```
Same here.

boomanaiden154 commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+        read_only=True,
+    )
+    url = result.stdout.strip()
+    match = re.search(r"github\.com[/:]([\w.-]+/[\w.-]+)", url)
```
Why do we need to pull out the repo name? This should always just be the upstream monorepo.

boomanaiden154 commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+
+    printer = Printer()
+    token = os.getenv("GITHUB_TOKEN")
+    default_prefix = "dev/"
```
`users/` given the monorepo will reject branches without a `users/` prefix?

joker-eph commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+        self.printer.print(f"On branch: {self.original_branch}")
+
+        try:
+            self._rebase_current_branch()
```
Why is this part of the flow?

joker-eph commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+                    "For stacks, the script must merge sequentially.",
+                    file=sys.stderr,
+                )
+                sys.exit(1)
```
Why is that? When pushing branches to upstream, you can cascade pull-requests to create your stack.

boomanaiden154 reviewed (commented):
I'm assuming you've tested this somewhere?

ilovepi commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+    def _get_commit_stack(self) -> List[str]:
+        """Gets the stack of commits between the current branch's HEAD and its merge base with upstream."""
+        target = f"{self.args.upstream_remote}/{self.args.base}"
+        merge_base_result = self._run_cmd(
```
IIRC it was to handle the case where main is ahead of your branch. IIRC I actually ran into this when prototyping.

ilovepi commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+    def _get_commit_details(self, commit_hash: str) -> tuple[str, str]:
+        """Gets the title and body of a commit."""
+        result = self._run_cmd(
+            ["git", "show", "-s", "--format=%s%n%n%b", commit_hash],
```
I prototyped this with a lot of 1 line commits (subject only) and some weirdly formatted bodies. having the subject line then `\n\n` and the rest being the body worked well.

ilovepi committed:
Rewrite script based on review feedback

After starting to address a few comments, I basically realized I needed
a new approach. I by and large kept the structure the same, but have
revised the library usage to ust use things from the standard library.

I've slightly modified some of the initial classes to be provide more
useful abstractions. There's probably room for more improvements, but
I wanted to keep this as simple as possible.

ilovepi commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+        self.printer.print(f"On branch: {self.original_branch}")
+
+        try:
+            self._rebase_current_branch()
```
Also, I found in my refactoring that if you don't have an accurate merge base w/ this approach, and you have a stack, you'll make redundant empty PRs if you don't say rebase your branch on top of main before trying to land again.

ilovepi replied:
I shored this up, and it seems to be working fine now that any PR landing always has a rebase operation following it. so the initial rebase is no longer necessary.

ilovepi committed:
Fix formatting

boomanaiden154 commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+    parser = argparse.ArgumentParser(
+        description="Create and land a stack of Pull Requests."
+    )
+    GITHUB_REMOTE_NAME = "origin"
```
I always clone `https://github.com/llvm/llvm-project`, so `origin` points to upstream and then I have `fork` as a remote to my fork.

Not sure what's more common though.

ilovepi committed:
Use API for branch deletion

ilovepi committed:
Use askpass over embeddign the token in the URL

ilovepi committed:
Fix git remote operations no longer using the token

Also fix up some comments

ilovepi committed:
Simplify rebae logic to keep base branch up to date

ilovepi commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,653 @@
+            self.printer.print("Cleaning up temporary local branches...")
+            self._run_cmd(["git", "branch", "-D"] + self.created_branches)
+            self.printer.print("Cleaning up temporary remote branches...")
+            self._run_cmd(
```
I don't think the delete operation ever happens when we have automerge enabled now. Do double check my logic, but when I tested it, that didn't happen, and I didn't see it in the verbose logging. The repo where I tested it didn't have much in the way of CI though, so its was just some linting. Still, I didn't see it retry or anything under `--verbose` when I tried it out last.

ilovepi committed:
Remove initial rebase

This isn't needed anymore, since we reliably rebase after every PR lands.

ilovepi committed:
Prefer %B over existing pattern

This was less trouble to adapt than I thought.

ilovepi committed:
Move git_env setup into a helper method

boomanaiden154 reviewed (commented):
I think this is pretty close.

ilovepi committed:
Fix typing import and restore FIXME

ilovepi committed:
Use inline helper over incorrect GIT_ASKPASS usage

ilovepi committed:
Simplify getting commits via rev-list

ilovepi committed:
Name groups for clarity

ilovepi committed:
Remove command wrapper method

ilovepi commented on llvm/utils/llvm_push_pr.py:
```diff
@@ -0,0 +1,696 @@
+        git_env["GIT_TERMINAL_PROMPT"] = "0"
+        return git_env
+
+    def _run_cmd(
```
done.

ilovepi committed:
Fix formatting

boomanaiden154 reviewed (approved):
LGTM at this point.

It would be good to add testing for this, but doing that properly requires mocking the Github API and that would be non-trivial to do initially and also maintain.

Let's get this shipped so we can start iterating on it in tree.

ilovepi committed:
Add documentation file

joker-eph reviewed (approved):
LG, let's mature this in tree, and move the doc to the LLVM development guide if/when we make PR mandatory and this replaces the current direct pushes to main.

ilovepi committed:
Rename files
